src/mito2/src/manifest/manager.rs

```diff
@@ -108,77 +189,83 @@ impl RegionManifestManagerInner {
-            manifest_builder.apply_change(RegionChange { metadata });
+            debug!("No region manifest in {}", options.manifest_dir);
+            return Ok(None);
         }
 
         let manifest = manifest_builder.try_build()?;
-        debug!("Recovered region manifest: {:?}", manifest);
-        let version = manifest.version.manifest_version;
+        debug!(
+            "Recovered region manifest from {}, manifest: {:?}",
+            options.manifest_dir, manifest
+        );
+        let version = manifest.manifest_version;
 
         // todo: start gc task
 
-        Ok(Self {
+        Ok(Some(Self {
             store,
             options,
-            version: AtomicManifestVersion::new(version),
-            manifest: ArcSwap::new(Arc::new(manifest)),
-        })
+            last_version: version,
+            manifest: Arc::new(manifest),
+        }))
     }
 
-    pub async fn stop(&self) -> Result<()> {
+    async fn stop(&mut self) -> Result<()> {
         // todo: stop gc task
         Ok(())
     }
 
-    pub async fn update(&self, action_list: RegionMetaActionList) -> Result<ManifestVersion> {
-        let version = self.inc_version();
-
+    async fn update(&mut self, action_list: RegionMetaActionList) -> Result<ManifestVersion> {
+        let version = self.increase_version();
         self.store.save(version, &action_list.encode()?).await?;
 
         let mut manifest_builder =
-            RegionManifestBuilder::with_checkpoint(Some(self.manifest.load().as_ref().clone()));
+            RegionManifestBuilder::with_checkpoint(Some(self.manifest.as_ref().clone()));
         for action in action_list.actions {
             match action {
                 RegionMetaAction::Change(action) => {
-                    manifest_builder.apply_change(action);
+                    manifest_builder.apply_change(version, action);
                 }
                 RegionMetaAction::Edit(action) => {
                     manifest_builder.apply_edit(version, action);
                 }
                 RegionMetaAction::Remove(_) | RegionMetaAction::Protocol(_) => {
-                    debug!("Unhandled action: {:?}", action);
+                    debug!(
+                        "Unhandled action for region {}, action: {:?}",
+                        self.manifest.metadata.region_id, action
+                    );
                 }
             }
         }
         let new_manifest = manifest_builder.try_build()?;
-        self.manifest.store(Arc::new(new_manifest));
+        self.manifest = Arc::new(new_manifest);
 
         Ok(version)
     }
 }
 
 impl RegionManifestManagerInner {
-    fn inc_version(&self) -> ManifestVersion {
-        self.version.fetch_add(1, Ordering::Relaxed)
+    /// Increases last version and returns the increased version.
+    fn increase_version(&mut self) -> ManifestVersion {
+        self.last_version += 1;
+        self.last_version
```

Comment:
Is there any consequent of returning the increased version instead?

Reply:
I keep `last_version` instead of `next_version` in the manager for convenience: I don't have to manually increase the version when I open or create the manager. Since it is `last_version`, I need to bump the version before returning.